src/ttsim/interface_dag_elements/results.py

```diff
@@ -14,29 +15,36 @@
 if TYPE_CHECKING:
     from ttsim.typing import (
         FlatData,
+        IntColumn,
         NestedData,
         NestedStrings,
         QNameData,
     )
 
 
 @interface_function()
-def tree(raw_results__combined: QNameData, input_data__flat: FlatData) -> NestedData:
-    """The combined results as a tree.
+def tree(
+    raw_results__columns: QNameData,
+    raw_results__params: QNameData,
+    raw_results__from_input_data: QNameData,
+    input_data__sort_indices: IntColumn,
+) -> NestedData:
+    """The combined results as a tree with original row order restored."""
+    restore_order = numpy.empty(len(input_data__sort_indices), dtype=int)
+    restore_order[input_data__sort_indices] = numpy.arange(
+        len(input_data__sort_indices)
+    )
 
-    The transformed id's are converted back to their original values.
+    def reorder_arrays(v: object) -> object:
+        return v[restore_order] if hasattr(v, "shape") and v.ndim > 0 else v  # type: ignore[index, attr-defined]
 
-    """
-    out = {}
-    for k in raw_results__combined:
-        path = dt.tree_path_from_qname(k)
-        if path in input_data__flat and (
-            path[-1].endswith("_id") or path[-1].startswith("p_id_")
-        ):
-            out[k] = input_data__flat[path]
-        else:
-            out[k] = raw_results__combined[k]
-    return dt.unflatten_from_qnames(out)
+    return dt.unflatten_from_qnames(
+        {
+            **raw_results__params,
+            **{k: reorder_arrays(v) for k, v in raw_results__from_input_data.items()},
```

Comment:
Wait, why do we need to reorder the input data? They should have never changed the order in the first place, I thought? 

Can you double-check by adding some column from the input data to the tt_targets of some end-to-end tests?

Reply:
Sorry, strike that! 

You are completely right, ofc, confused myself that `raw_results__from_input_data` uses `processed_data` as an input. 

Terrible naming though.

@MImmesberger, should we rename to `from_processed_data` ? It is not really user-facing, so keeping the internals straight seems the greater good relative to using the intuitive name to the outside.